File: all/226_invert_binary_tree/sol_preorder.py
from typing import Optional


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def invertTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        def invertTreeByPreorderTraversal(root: Optional[TreeNode]) -> None:
            if not root:
                return

            # Swap the left subtree and the right subtree.
            (root.left), (root.right) = (root.right), (root.left)

            # Continue the same procedures on the left subtree and the right
            # subtree.
            invertTreeByPreorderTraversal(root.left)
            invertTreeByPreorderTraversal(root.right)

        invertTreeByPreorderTraversal(root)

        return root

File: all/226_invert_binary_tree/sol_preorder.c
#include <stddef.h>

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */
void invertTreeByPreorderTraversal(struct TreeNode * const root) {
    struct TreeNode *original_left_subtree_ptr = NULL;

    if (NULL == root) {
        return;
    }

    /* Swap the left subtree and right subtree. */
    original_left_subtree_ptr = root->left;
    root->left = (root->right);
    root->right = original_left_subtree_ptr;

    /* Continue the same procedure on the left subtree and the right subtree. */
    invertTreeByPreorderTraversal(root->left);
    invertTreeByPreorderTraversal(root->right);
}

struct TreeNode* invertTree(struct TreeNode* root) {
    /*
     * User recursive preorder traversal to swap the left subtree and
     * right subtree of a given root node.
     */
    invertTreeByPreorderTraversal(root);

    return root;
}
